source/Halibut/Transport/TcpClientManager.cs

```diff
@@ -58,37 +60,47 @@ public IReadOnlyCollection<TcpClient> GetActiveClients(string thumbprint)
 
         public void RemoveClient(TcpClient client)
         {
-            lock (activeClients)
+            lock (syncLock)
             {
                 foreach(var thumbprintClientsPair in activeClients)
                 {
                     if (thumbprintClientsPair.Value.Contains(client))
+                    {
                         thumbprintClientsPair.Value.Remove(client);
+                    }
                 }
 
                 var thumbprintsWithNoClients = activeClients
                     .Where(x => x.Value.Count == 0)
                     .Select(x => x.Key)
                     .ToArray();
+
                 foreach (var thumbprint in thumbprintsWithNoClients)
+                {
                     activeClients.Remove(thumbprint);
+                }
             }
         }
 
         public void Dispose()
         {
-            var clients = activeClients?.ToArray();
-            activeClients?.Clear();
-
-            if (clients == null || !clients.Any())
-            {
-                return;
-            }
-            
-            foreach (var client in clients)
+            lock (syncLock)
             {
-                if (client.Value?.Any() == true)
+                var clients = activeClients?.ToArray();
+                activeClients = null;
```

Comment:
This will help us identify what is modifying this list during disposal / what we are not correctly stopping in Halibut etc.